### PR TITLE
Add additional limits to fields in Receipts, Individual contributions, and Disbursements datatable

### LINKED ISSUE #3779: Add 10 item limitation to filters on schedule a, b and e datatables
**What we're after:** Following a brief outage on 5/21, we need to implement front-end filter limitations on Schedule A, B and E for: **name, ZIP (limit 5), employer, occupation, and city.** 

Datatables where we're making the change: [All receipts](https://www.fec.gov/data/receipts/), [Individual contributions](https://www.fec.gov/data/receipts/individual-contributions/), [Disbursements](https://www.fec.gov/data/disbursements/), [Independent expenditures](https://www.fec.gov/data/independent-expenditures/)

**All receipts and individual contributions datatables:** 
- [x] Limit 10 for:
- Name - `contributor_name`
- Zip - `contributor_zip`
- City - `contributor_city`
- Occupation - `contributor_employer`
- Employer - `contributor_occupation`

**Disbursements datatable:** 
- [x] Limit 10 for:
- Recipient name or ID - `recipient_name` (takes ID or name)
- City  - `recipient_city`

**Independent expenditures datatable:** 
- [x] Limit 10 for:
- Candidate name or ID (`candidate_id`)

**Related ticket:**
- https://github.com/fecgov/openfec/issues/4118
- https://github.com/fecgov/openFEC/issues/3902

**Completion criteria:**
- [x] Implement 10 item limitation for the aforementioned filters on the above datatables. Mimic the 10 committee limitation interaction. Design work here: fecgov/fec-cms#3272

Future work:
- Figure out how to log that the limit is triggered in the server logs

### PR DESCRIPTION
**Pull request by patphongs**
## Summary

- Resolves #3779 
_Adds additional item limits for the following fields:_

**Individual contributions and Receipts datatable**

Recipient name or ID - committee_id
Name - contributor_name
Zip - contributor_zip
City - contributor_city
Occupation - contributor_employer
Employer - contributor_occupation

**Disbursements datatable**
Spender name or ID - committee_id
Recipient name or ID - recipient_name
City - recipient_city

**Independent expenditures datatable**
Spender name or ID - committee_id
Candidate name or ID - candidate_id

## Impacted areas of the application

List general components of the application that this PR will affect:

- Receipts datatable
- Individual contributions datatable
- Disbursements datatable
- Independent expenditures datatable

### Receipts and individual contributions datatable
receipts-datatable-limit-10

### Disbursements datatable
disbursements-datatable-limit-10

### Independent expenditures datatable
ie-datatable-limit-10

## Related PRs

List related PRs against other branches:

branch | PR
------ | ------
feature/3367-cmte-id-limit-msg | Datatables handle >10 committee_ids https://github.com/fecgov/fec-cms/pull/3380

### Receipts and individual contributions datatable 
You can use these same URLs to test the individual contributions errors. Just replace the `receipts` with `individual-contributions` in the URL.

**Over 10 committee IDs:**
http://localhost:8000/data/receipts/?data_type=processed&committee_id=C00213850&committee_id=C00222273&committee_id=C00222612&committee_id=C00222760&committee_id=C00222778&committee_id=C00223446&committee_id=C00223735&committee_id=C00223743&committee_id=C00223925&committee_id=C00224048&committee_id=C00452706&two_year_transaction_period=2020&min_date=01%2F01%2F2019&max_date=12%2F31%2F2020

**Over 10 contributor names:**
http://localhost:8000/data/receipts/?data_type=processed&contributor_name=blaine&contributor_name=connie&contributor_name=demi&contributor_name=elephant&contributor_name=flanders&contributor_name=gilbert&contributor_name=hellen&contributor_name=isabella&contributor_name=janice&contributor_name=kimberly&contributor_name=latrice&two_year_transaction_period=2020&min_date=01%2F01%2F2019&max_date=12%2F31%2F2020

**Over 10 cities:**
http://localhost:8000/data/receipts/?data_type=processed&contributor_city=albany&contributor_city=belmont&contributor_city=blacksburg&contributor_city=cascades&contributor_city=cincinati&contributor_city=clarksburg&contributor_city=dublin&contributor_city=houston&contributor_city=raleigh&contributor_city=richmond&contributor_city=seattle&two_year_transaction_period=2020&min_date=01%2F01%2F2019&max_date=12%2F31%2F2020

**Over 10 zip codes:**
http://localhost:8000/data/receipts/?data_type=processed&contributor_zip=06489&contributor_zip=07032&contributor_zip=07109&contributor_zip=10701&contributor_zip=11720&contributor_zip=30213&contributor_zip=31021&contributor_zip=37659&contributor_zip=38801&contributor_zip=43560&contributor_zip=55406&two_year_transaction_period=2020&min_date=01%2F01%2F2019&max_date=12%2F31%2F2020

**Over 10 occupations:** 
http://localhost:8000/data/receipts/?data_type=processed&contributor_occupation=actor&contributor_occupation=analyst&contributor_occupation=ceo&contributor_occupation=doctor&contributor_occupation=engineer&contributor_occupation=manager&contributor_occupation=musician&contributor_occupation=programmer&contributor_occupation=psychologist&contributor_occupation=self+employed&contributor_occupation=teacher&two_year_transaction_period=2020&min_date=01%2F01%2F2019&max_date=12%2F31%2F2020

**Over 10 employers:** 
http://localhost:8000/data/receipts/?data_type=processed&contributor_employer=amazon&contributor_employer=apple&contributor_employer=autobody+shop&contributor_employer=borrow+group&contributor_employer=brita&contributor_employer=corporation&contributor_employer=dominos&contributor_employer=geico&contributor_employer=hospital&contributor_employer=mozilla&contributor_employer=restaurant&min_date=01%2F01%2F2019&max_date=12%2F31%2F2020

### Disbursements datatable

**Over 10 committees:**
http://localhost:8000/data/disbursements/?data_type=processed&committee_id=C00213850&committee_id=C00222273&committee_id=C00222612&committee_id=C00222760&committee_id=C00222778&committee_id=C00223446&committee_id=C00223735&committee_id=C00223743&committee_id=C00223925&committee_id=C00224048&committee_id=C00452706&two_year_transaction_period=2020&min_date=01%2F01%2F2019&max_date=12%2F31%2F2020

**Over 10 recipient names:**
http://localhost:8000/data/disbursements/?data_type=processed&recipient_name=blaine&recipient_name=connie&recipient_name=demi&recipient_name=elephant&recipient_name=flanders&recipient_name=gilbert&recipient_name=hellen&recipient_name=isabella&recipient_name=janice&recipient_name=kimberly&recipient_name=latrice&two_year_transaction_period=2020&min_date=01%2F01%2F2019&max_date=12%2F31%2F2020

**Over 10 cities:**
http://localhost:8000/data/disbursements/?data_type=processed&two_year_transaction_period=2020&min_date=01%2F01%2F2019&max_date=12%2F31%2F2020&recipient_city=seattle&recipient_city=richmond&recipient_city=raleigh&recipient_city=houston&recipient_city=dublin&recipient_city=clarksburg&recipient_city=cincinati&recipient_city=cascades&recipient_city=blacksburg&recipient_city=belmont&recipient_city=albany

### Independent expenditures datatable
**Over 10 committees:**
http://localhost:8000/data/independent-expenditures/?data_type=processed&is_notice=true&committee_id=C00213850&committee_id=C00222273&committee_id=C00222612&committee_id=C00222760&committee_id=C00222778&committee_id=C00223446&committee_id=C00223735&committee_id=C00223743&committee_id=C00223925&committee_id=C00224048&committee_id=C00452706

**Over 10 candidates:**
http://localhost:8000/data/independent-expenditures/?data_type=processed&is_notice=true&most_recent=true&candidate_id=H8CA05035&candidate_id=H8WI01024&candidate_id=P00000455&candidate_id=P00003335&candidate_id=P00003392&candidate_id=P00010298&candidate_id=P00014530&candidate_id=P20000642&candidate_id=P80003338&candidate_id=S2GA00019&candidate_id=S4AK00099

Include any information that may be helpful to the reviewer(s).
_This might include:_

- links to sample pages to test
- Any local environmental setup that is unusual (env vars, API version to point to, etc)

____

### CODE REVIEW

File: fec/fec/static/js/modules/tables.js [hunk 1]
```diff
@@ -645,7 +645,7 @@ DataTable.prototype.fetch = function(data, callback) {
       candidate_id: `You&#39;re trying to search more than ${MAX_FILTERS} candidates. Narrow your search to ${MAX_FILTERS} or fewer candidates.`,
       contributor_name: `You&#39;re trying to search more than ${MAX_FILTERS} contributors. Narrow your search to ${MAX_FILTERS} or fewer contributors.`,
       recipient_name: `You&#39;re trying to search more than ${MAX_FILTERS} recipients. Narrow your search to ${MAX_FILTERS} or fewer recipients.`,
-      contributor_zip: `You&#39;re trying to search more than ${MAX_FILTERS} zip codes. Narrow your search to ${MAX_FILTERS} or fewer zip codes.`,
+      contributor_zip: `You&#39;re trying to search more than ${MAX_FILTERS} ZIP codes. Narrow your search to ${MAX_FILTERS} or fewer ZIP codes.`,
```
Comment (patphongs) [inline fec/fec/static/js/modules/tables.js:648]: Thanks @dorothyyeager for identifying the casing issue for ZIP codes. This has been fixed.

**Comment by lbeaufort** [comment]
@rfultz thanks for your careful review and documentation! 

@patphongs I think the "you're searching across more than one 2-year time period" is a result of the first 400-error call that the standard load page always gets - maybe the 2-year period is being applied on the second call, but with the new validation, the second call never happens?

Standard call:
http://localhost:8000/data/receipts/?data_type=processed&two_year_transaction_period=2020&min_date=01%2F01%2F2019&max_date=12%2F31%2F2020
Screen Shot 2020-05-29 at 9 33 48 AM

File: fec/fec/static/js/modules/tables.js (hunk 1 above)
Comment (dorothyyeager) [reply fec/fec/static/js/modules/tables.js:648]: Thanks @patphongs !

**Comment by JonellaCulmer** [comment]
@patphongs I'm going to wait to do a more thorough review after Robert and Laura's comments are addressed. One comment though about the background colors behind some of the error messages. Some have a lighter background and others don't. I prefer the lighter background, but we haven't done that consistently. I can open a follow-up ticket to make that happen across all error messages, if you'd like to add it to this work.

**Comment by patphongs** [comment]
@rfultz @lbeaufort @JonellaCulmer Thanks for the detailed review! Here's my comments:

## Addressed issues

1. Removed the 2 year transaction period error if there's a greater than 10 error response on a limited field. This is occurring due to the `fetchError` state, in which I believe it's triggering due to the processed and raw toggles. I haven't been able to figure out how to get rid of that 400 yet.

2. Expand any accordion that has the error message in it.

3. Made all the checkboxes that are on the dropdown filters have an unsuccessful `X` mark next to it if the error message exists. I think this is important for the user to know that none of the filters will be successful until they address the over 10 limit error. @JonellaCulmer What do you think about this?

Screen Shot 2020-06-01 at 2 07 58 PM 

## Unaddressed issues

Two things I haven't addressed that could be floated into a new ticket: 

1. `NaN results` display when checking and unchecking. Need to make a ticket once this work is merged in.

2. Background color for error message does not apply the light grey like in the accordion messages. Need to figure out why the `.message` class is not applying the light grey color.

File: fec/fec/tests/js/tables.js [hunk 2]
```diff
@@ -483,71 +483,107 @@ describe('data table', function() {
...
-        committee_id: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
+        committee_id: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11],
+        candidate_id: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11],
+        contributor_name: ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k'],
+        recipient_name: ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k'],
+        contributor_zip: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11],
+        contributor_city: ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k'],
+        recipient_city: ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k'],
+        contributor_employer: ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k'],
+        contributor_occupation: ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k'],
```
[the remark this answers is not in the document]
Comment (patphongs) [reply fec/fec/tests/js/tables.js]: Thanks @rfultz, I've consolidated these into variables

**Comment by JonellaCulmer** [comment]
@patphongs My notes: 
- Year error appearing when it shouldn't because it's a single year and no error for committees shows when there are 11 committees total. No data showing in the table and no overlay to suggest it's "thinking." 
- Filter on the left column is acting like an overall click instead of removing when 11 committees are selected. Clicking the red X acts as check, doesn't remove. 
- If there's an error in one filter group, every applied filter on the left column lights up red, even if there's no problem with the other filter groups. Only that group with 11 items should light up, not all with filters applied. See screenshot below:
Screen Shot 2020-06-01 at 4 38 07 PM

File: fec/data/templates/partials/individual-contributions-filter.jinja [hunk 3]
```diff
@@ -14,13 +14,13 @@ Filter individual contributions
 
 {% block filters %}
 <div class="filters__inner">
-  {{ typeahead.field('committee_id', 'Recipient Name or ID') }}
+  {{ typeahead.field('committee_id', 'Recipient Name or ID', helper_text='Limit 10 recipient names or IDs') }}
```
Comment (rfultz) [inline fec/data/templates/partials/individual-contributions-filter.jinja:17]: Should we change all of the `helper_text='` to `helper_text = '` for consistency?

**Comment by rfultz** [comment]
For disbursements datatable Over 10 cities, when I load the link, the time period and date range loading animations are playing. I don't know how high that is on the priority list. Doesn't seem like a reason to hold up the filter limit.

I don't know that there's anything to do about it (i.e. I think it's a browser issue and limited): when I open the last few links in the main body of this PR in new tabs, when I switch to those tabs, the red outline doesn't appear until I interact with them or reload. I suspect it's because of browsers' restrictions against JavaScript running in background tabs

**Comment by rfultz** [comment]
Thank you, Pat, for wrestling all the details and possible use and error cases for us!

File: fec/data/templates/partials/individual-contributions-filter.jinja (hunk 3 above)
Comment (patphongs) [reply fec/data/templates/partials/individual-contributions-filter.jinja]: Thanks, added for consistency

**Comment by patphongs** [comment]
Thanks @rfultz, please see my comments below:

> For disbursements datatable Over 10 cities, when I load the link, the time period and date range loading animations are playing. I don't know how high that is on the priority list. Doesn't seem like a reason to hold up the filter limit.

The loading icon is kind of a trade off. Originally, I had it where all of the filter checkboxes will have a red `X` next to the labels. But we decided that was too confusing if there was no error on that field, so I changed it back. I guess we should think about what we would want to be there if no loading is happening? No loading symbol or indicator at all?
 
> I don't know that there's anything to do about it (i.e. I think it's a browser issue and limited): when I open the last few links in the main body of this PR in new tabs, when I switch to those tabs, the red outline doesn't appear until I interact with them or reload. I suspect it's because of browsers' restrictions against JavaScript running in background tabs

The red `X` and red boxes should appear if there's an error that exists for that field. I tried it locally and it worked. Do you think it's caching?

**Pull request merged**